lib/chef/knife/opc_user_delete.rb

```diff
@@ -19,27 +19,145 @@
 
 module Opc
   class OpcUserDelete < Chef::Knife
+  class DeleteFailed < StandardError ; end
     category "OPSCODE PRIVATE CHEF ORGANIZATION MANAGEMENT"
-    banner "knife opc user delete USERNAME [-d]"
+    banner "knife opc user delete USERNAME [-d] [-R]"
 
     option :no_disassociate_user,
-    :long => "--no-disassociate-user",
-    :short => "-d",
-    :description => "Don't disassociate the user first"
+      :long => "--no-disassociate-user",
+      :short => "-d",
+      :description => "Don't disassociate the user first"
 
+    option :remove_from_admin_groups,
+      long:  "--remove-from-admin-groups",
+      short:  "-R",
+      description: "If the user is a member of any org admin groups, attempt to remove from those groups. Ignored if --no-disassociate-user is set."
+
+    attr_reader :username
     include Chef::Mixin::RootRestv0
 
+    deps do
+      require 'chef/org'
+      require 'chef/org/group_operations'
+    end
+
     def run
-      username = @name_args[0]
+      @username = @name_args[0]
+      admin_memberships = []
+      unremovable_memberships = []
+
       ui.confirm "Do you want to delete the user #{username}"
-      unless config[:no_disassociate_user]
-         orgs =  root_rest.get("users/#{username}/organizations")
-         org_names = orgs.map {|o| o['organization']['name']}
-         org_names.each do |org|
-            ui.output root_rest.delete("organizations/#{org}/users/#{username}")
-         end
+
+      if config[:no_disassociate_user]
+        delete_user(username)
+        return
+      end
+
+      ui.output("Checking organization memberships...")
+      orgs = org_memberships(username)
+
+      if orgs.length > 0
+        ui.output("Checking admin group memberships for #{orgs.length} org(s).")
+        admin_memberships, unremovable_memberships = admin_group_memberships(orgs, username)
+      end
+
+      if admin_memberships.empty?
```

Comment:
Both sides of this branch end with the same calls to disassociate_user and delete_user.  If you reverse this conditional (`if !admin_memberships.empty?`) you can then have this focus on just what is different, namely the case where we want to remove the user from the admin group.

It also feels a bit odd to me to introduce an exception just to rescue it in the same function and print an error.  Since we want to exit with a non-zero exit status in these cases anyway, maybe this is a case where it would be cleaner to early exit?

```ruby
if !admin_memberships.empty?
  if !config[:remove_from_admin_groups]
    # We haven't been given -R, exit with error
    exit_with_error_admin_group_member!(username,  admin_memberships))
  elsif !unremovable_memberships.empty?
    # We've been given -R but we know the delete will fail for these orgs
    exit_with_error_cant_remove_admin_membership!(username, unremovable_memberships))
  else
   # Try to remove this user from all the admin groups
    remove_from_admin_groups(admin_memberships, username)
  end
end

dissociate_user(orgs, username)
delete_user(username)
```